### ISSUE TITLE
skore.cross_validate 

### ISSUE THREAD

**Issue by MarieS-WiMLDS** [probabl-ai/skore#383]
# Background

https://www.notion.so/probabl/2024-09-20-Workshop-with-Ga-l-107ef76d36b980fdbba2c6036db91a35  

# Long term goals

We want to help by: 
- allowing easy comparison 
- building report of the results once frozen

# What do we want to save?

The `cross_validate` function output a dict with 6 keys according to [scikit-learn doc](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_validate.html). Each value is an array of length n_splits.  

We want to do a skore wrapping around the sklearn function: it should behave the same way, only adding some storage and graph computing if deemed relevant.  
To reach our goals, and to provide help to the user, I need to remember the context around when the function `cross_validate` is launched.  
In the long term, I want to be able to compare the various runs of cross validate, with something that could look like this:  
![WhatsApp Image 2024-09-23 at 18 05 23](https://github.com/user-attachments/assets/e0ff50db-c56e-4bb6-9067-4f31ad750387)

It would be possible to filter on some elements, to select only one kind of estimator for instance. We would also forbid comparison on different scorings (it doesn't make sense to compare f1 with accuracy).

**Comment by tuscland** [comment]
2024-09-30 -- Meeting report with @glemaitre and @ogrisel

## Proposed Enhancements:

1. **Storage of Probability and Decision Function:**
   - Store predict_proba and decision_function results to enable:
     - Calculation of additional metrics later
     - Interactive threshold adjustment without re-running cross-validation

2. **Precision/Recall Display:**
   - Show precision/recall metrics
   - Explicitly state the threshold used (typically 0.5, but could be something else)

3. **Business Metric Integration:**
   - If a user registers a business metric (via a registry?), recalculate it as well
   - Note: Scikit-learn scorers don't allow choosing scorers after the fact

4. **Dataset Statistics:**
   - Include sufficient statistics of a dataset in the cross-validation result

5. **Threshold Tuning:**
   - Using raw predictions (cross-validation results):
     - Extract code from TTCC (trivial)
     - Optimize business metric between a minimum and maximum threshold
     - Calculate average for each split of the cross-validation
   - Reference: [Scikit-learn Cost-Sensitive Learning Example](https://scikit-learn.org/dev/auto_examples/model_selection/plot_cost_sensitive_learning.html#consideration-regarding-model-refitting-and-cross-validation)

These enhancements aim to improve the flexibility and utility of our cross-validation process, allowing for more detailed analysis and optimization of model performance.

**Comment by MarieS-WiMLDS** [comment]
Additional notes from discussion with Auguste:
- the graph names should be unique as we write them in the same `skore` database than was it created by the user. Maybe start with `_` and forbid such names for variables created by the user?
- the functions to create the plots should be private, for now no reason to let the user access it directly yet.
- We must output the same thing to avoid breaking any workflow, the change can be done smoothly. It means removing the additional outputs produced by the addition of new scoring functions.

**Comment by tuscland** [comment]
> the graph names should be unique as we write them in the same skore database than was it created by the user. Maybe start with _ and forbid such names for variables created by the user?

Without going into the details of the feature (because I don't understand it), I would not use conventions (underscore) to implement this.

**Comment by MarieS-WiMLDS** [comment]
# Improvements

## Pain point not adressed so far

I train my model and cv it.  
Two months later, I see in despair that I forgot to calculate this business metric (a Python function) that I wrote two days ago.  
How do I measure this metric with my past trainings?  
-> This is a great feature!  

## Blocker 

The `cross_validate` function from scikit-learn doesn't output the y_proba. Plus, we shouldn't rely on the estimators storage because it would require to restore the exact same environment (and it could change meanwhile, e.g. sklearn version). 

## Potential solution

Inside our cross_validate function, using the estimators & split indices, compute y_proba, and then store it.

**Comment by augustebaum** [comment]
From @MarieS-WiMLDS at https://github.com/probabl-ai/skore/pull/443#discussion_r1799629341

> At first, we will ask the user to say if it's comparable, otherwise we will consider that they are always comparable inside the same project. Each run will be named `run_{n}`, and will have a "info" chart that can be displayed on hover, containing: the estimator name, number of cv splits, id of the data used (to be defined yet, displaying an hash is not very user-friendly).
